src/hotspot/share/utilities/exceptions.cpp

```diff
@@ -160,8 +160,17 @@ void Exceptions::_throw(JavaThread* thread, const char* file, int line, Handle h
     return;
   }
 
-  if (h_exception->is_a(vmClasses::OutOfMemoryError_klass())) {
-    count_out_of_memory_exceptions(h_exception);
+  if (h_exception->is_a(vmClasses::VirtualMachineError_klass())) {
+    // Remove the ScopedValue cache in case we got a virtual machine
+    // Error while we were trying to manipulate ScopedValue bindings.
+    thread->set_scopedValueCache(NULL);
```

Comment:
I am see this pattern repeat quite often:
```
    thread->set_scopedValueCache(NULL);
    oop threadObj = thread->vthread();
    assert(threadOop != NULL, "must be"); // <--- sometimes
    java_lang_Thread::clear_scopedValueBindings(threadObj);
```
Encapsulate in a method on the `JavaThread` class?

Reply:
That sounds good.

Reply:
Done.